### ISSUE TITLE
[REFACTOR] amp 적용 등 학습 효율 향상을 위한 것들

### ISSUE THREAD

**Issue by 2018007956** [boostcampaitech6/level1-imageclassification-cv-14#59]
## Background
- 2023-12-19 이동빈님의 오피스아워를 듣고 얻은 학습 효율을 높일 수 있는 꿀팀들 적용해보기

## To do
- [x] Mixed Precision Training
    - 일부 FP 16 변환하여 FP32&FP16 함께 사용 → 연산속도, 메모리 사용량 줄어듬
    - `torch.cuda.amp.autocast()`
- [x] gradients = None rather than 0
    - PyTorch 1.7 이상에서 set_to_none parameter 제공
- [x] CPU → GPU 올리는 과정을 비동기로 처리
    - batch 커지거나 데이터 전처리 많아져서 데이터를 GPU에 올리는데 시간 많이 드는 경우 효과적
    - `to(device, non_blocking=True)`
- Fast prefetcher
    - pytorch에는 cuda.Stream() 과 같이 CPU → GPU 흐름에 대한 기능을 제공
    - Augmentation 연산을 GPU에서 할 수 있도록 함
    - to tensor, normalize 를 GPU에서 처리할 수 있도록 transforms 함수에는 resize, ToNumpy()만 넣어놓고, 나중 PrefetchLoader class에 추가하여 처리
    - 기존 DataLoader를 PrefetchLoader로 감싸줌
- 위를 모두 적용했을 때 최종 결과
![image](https://github.com/boostcampaitech6/level1-imageclassification-cv-14/assets/48304130/d866d9f3-ebd0-4b83-8c0d-668ac9486349)
